45.md

```diff
@@ -0,0 +1,209 @@
+NIP-45
+======
+
+Nostr Marketplace (for resilient marketplaces)
+-----------------------------------
+
+`draft` `optional` `author:fiatjaf` `author:benarc`  `author:motorina0` `author:talvasconcelos` 
+
+> Based on https://github.com/lnbits/Diagon-Alley
+
+## Terms
+
+- `merchant` - seller of products with NOSTR key-pair
+- `customer` - buyer of products with NOSTR key-pair
+- `product` - item for sale by the `merchant`
+- `stall` - list of products controlled by `merchant` (a `merchant` can have multiple stalls)
+- `marketplace` - clientside software for searching `stalls` and purchasing `products`
+
+## Nostr Marketplace Clients
+
+### Merchant admin
+
+Where the `merchant` creates, updates and deletes `stalls` and `products`, as well as where they manage sales, payments and communication with `customers`.
+
+The `merchant` admin software can be purely clientside, but for `convenience` and uptime, implementations will likely have a server client listening for NOSTR events.
+
+### Marketplace
+
+`Marketplace` software should be entirely clientside, either as a stand-alone app, or as a purely frontend webpage. A `customer` subscribes to different merchant NOSTR public keys, and those `merchants` `stalls` and `products` become listed and searchable. The marketplace client is like any other ecommerce site, with basket and checkout. `Marketplaces` may also wish to include a `customer` support area for direct message communication with `merchants`.
+
+## `Merchant` publishing/updating products (event)
+
+A merchant can publish these events:
+| Kind    |                  | Description                                                                                                   | NIP                                     |
+|---------|------------------|---------------------------------------------------------------------------------------------------------------|-----------------------------------------|
+| `0    ` | `set_meta`       | The merchant description (similar with any `nostr` public key).                                               | [NIP01       ](https://github.com/nostr-protocol/nips/blob/master/01.md)                            |
+| `30017` | `set_stall`      | Create or update a stall.                                                                                     | [NIP33](https://github.com/nostr-protocol/nips/blob/master/33.md) (Parameterized Replaceable Event) |
+| `30018` | `set_product`    | Create or update a product.                                                                                   | [NIP33](https://github.com/nostr-protocol/nips/blob/master/33.md) (Parameterized Replaceable Event) |
+| `4    ` | `direct_message` | Communicate with the customer. The messages can be plain-text or JSON. | [NIP09](https://github.com/nostr-protocol/nips/blob/master/09.md)                                            |
```

Comment:
Please don't use the kind 4, use another number so that it doesn't get mixed with regular DM's in clients

Reply:
- this is the only `kind` that its encrypted
- it has the (privacy) advantage that it "looses in the crowd". Otherwise it would be much easier to identify orders.
- the key (`npub`) that a customer uses to buy goods might be different than the one used for social media
- some social media clients (like `hamstr`) have nice way of displaying JSON (in the form of tables)

Reply:
- Use the same encryption scheme 
- Stores and clients might use different keys anyway so this doesnt really add any noise 
- Yes
- Most don't and won't parse json messages in DM's

Reply:
> Stores and clients might use different keys anyway so this doesnt really add any noise to hide in

The noise is network-wide not key-specific. Although since the other kinds determine the pubkey of the marketplace, you are right, it is obvious the comms are for marketplace actions.

@motorina0 
The only way to make this work is to do what I did in https://github.com/lnurl/luds/pull/203 and have an extra interaction round so that the stall tells the customer which pubkey to communicate with.

Reply:
@Kukks 
- having a new `kind` for "typed/structured" DMs is also an option. See https://github.com/nostr-protocol/nips/issues/394#issuecomment-1483908687
- a more generic mechanism for changing the keys for DMs would also solve the privacy issue for markets. See https://github.com/nostr-protocol/nips/issues/393
   - it is not the complete solution, defense against track-by-relays to be added